Jenkinsfile

```
@@ -1,10 +1,14 @@
 #!groovy
 
+Random random = new Random() // Randomize which Jenkins version is selected for more testing
+def use_newer_jenkins = random.nextBoolean() // Use newer Jenkins on one build but slightly older on other
+
 // Test plugin compatibility to recommended configurations
-// Allow failing tests to retry execution
 subsetConfiguration = [ [ jdk: '8',  platform: 'windows', jenkins: null                      ],
-                        [ jdk: '8',  platform: 'linux',   jenkins: '2.164.1', javaLevel: '8' ],
-                        [ jdk: '11', platform: 'linux',   jenkins: '2.164.1', javaLevel: '8' ]
+                        // Compile with Java 8, test 2.204.6 or 2.222.1 depending on random use_newer_jenkins
+                        [ jdk: '8',  platform: 'linux',   jenkins: !use_newer_jenkins ? '2.204.6' : '2.222.1', javaLevel: '8' ],
+                        // Compile with Java 11, test the Jenkins version that Java 8 did *not* test
+                        [ jdk: '11', platform: 'linux',   jenkins:  use_newer_jenkins ? '2.204.6' : '2.222.1', javaLevel: '8' ]
                       ]
```

This doesn't make the builds very deterministic. And rebuilding something could cause it to fail or not depending on what version got selected.
A slightly better thing might be to add on  those stages instead?
```suggestion
                        // Compile with Java 8, 
                        [ jdk: '8',  platform: 'linux',   jenkins: '2.204.6' , javaLevel: '8' ],  
                        // // Compile with Java 11                      
                        [ jdk: '11', platform: 'linux',   jenkins:  '2.204.6', javaLevel: '8' ]
                      ]
//test 2.222.1 depending on random use_newer_jenkins
if (use_newer_jenkins) {
    subsetConfiguration << [ jdk: '8',  platform: 'linux',   jenkins: '2.222.1' , javaLevel: '8' ]
    subsetConfiguration << [ jdk: '11', platform: 'linux',   jenkins:  '2.222.1', javaLevel: '8' ]
}
```

---

Thanks for the suggestion, but I'm unwilling to expand the tested configurations beyond the three that are tested here.

The tests in the git plugin and the git client plugin are already more expensive to run than I believe is justified.  Their ability to detect bugs by running in even more configurations is quite low and the cost of time and compute resources on ci.jenkins.io is enough that it overshadows any gain from the tests.

We currently average 10+ pull requests open against the git plugin and the git client plugin.  Tests are executed against 10+ pull requests after a commit to the master branch the next time the repository is scanned .  With that many test runs after a commit to the master branch, the odds are very much in our favor that all combinations are executed at least once.

In a similar vein, my [Jenkins public test infrastructure](https://github.com/MarkEWaite/docker-lfs/tree/lts-with-plugins)  and its [private component](https://github.com/MarkEWaite/docker-private-lfs/tree/lts-with-plugins-add-credentials-and-nodes) also run the git client plugin and git plugin tests regularly.

If the tests were significantly cheaper (faster to run, less memory, less network, etc.) or if there were very few open pull requests, my approach might be different.

A similar approach is taken inside the tests themselves in many cases.  When a test needs a particular operation but does not care how the operation is performed (checkout or clone or tag or ...), random selectors are frequently used to choose an arbitrary implementation (CLI git, JGit, etc.).  That technique has helped to assure that the independent implementations of functionality are known to be reasonably compatible with each other, or at least that tests document the cases where they are incompatible.